src/main/java/de/zalando/aruha/nakadi/service/subscription/model/Partition.java

```diff
@@ -0,0 +1,124 @@
+package de.zalando.aruha.nakadi.service.subscription.model;
+
+import java.util.Collection;
+
+public class Partition {
+    public static class PartitionKey {
+        public final String topic;
+
+        public final String partition;
+
+        public PartitionKey(final String topic, final String partition) {
+            this.topic = topic;
+            this.partition = partition;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            final PartitionKey that = (PartitionKey) o;
+
+            return topic.equals(that.topic) && partition.equals(that.partition);
+
+        }
+
+        @Override
+        public int hashCode() {
+            int result = topic.hashCode();
+            result = 31 * result + partition.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "{" + topic + ':' + partition + '}';
+        }
+    }
+
+    public enum State {
+        UNASSIGNED,
+        REASSIGNING,
+        ASSIGNED,;
+    }
+
+    private final PartitionKey key;
+    private final String session;
+    private final String nextSession;
+    private final State state;
+
+    public Partition(final PartitionKey key, final String session, final String nextSession, final State state) {
+        this.key = key;
+        this.session = session;
+        this.nextSession = nextSession;
+        this.state = state;
+    }
+
+    public Partition toState(final State state, final String session, final String nextSession) {
+        return new Partition(key, session, nextSession, state);
+    }
+
+    public Partition moveToSessionId(final String sessionId, final Collection<String> existingSessionIds) {
+        switch (state) {
+            case UNASSIGNED:
+                return toState(State.ASSIGNED, sessionId, null);
+            case ASSIGNED:
+                if (sessionId.equals(this.session) || !existingSessionIds.contains(this.session)) {
+                    return toState(State.ASSIGNED, sessionId, null);
```

Comment:
I mean, why not return the same object?

In the case of the second condition (after `||`) I do understand that the current session no longer exists so it should be moved to reassigning.